tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,594 @@
+    const legendYLoc =
+        this._hostGridHeight * Math.ceil(this._yDimension / HOST_Y_STRIDE) +
+        HOST_TO_HOST_MARGIN;
+    this.drawLegend(svg, legendYLoc, colorScale);
+  },
+  /**
+   * Returns the size of host grid, including the host card size and the margin
+   * between two hosts.
+   */
+  getHostGridSize(stride: number): number {
+    return HOST_TO_CHIP_MARGIN * 2 + CHIP_TO_CHIP_MARGIN * (stride - 1) +
+        CHIP_GRID_SIZE * stride + HOST_TO_HOST_MARGIN;
+  },
+  /**
+   * Returns the x-axis location for the xChip'th chip of the xHost'th host.
+   */
+  getChipXLoc: function(xHost: number, xChip: number): number {
+    return xHost * this._hostGridWidth + HOST_TO_CHIP_MARGIN +
+        xChip * (CHIP_GRID_SIZE + CHIP_TO_CHIP_MARGIN);
+  },
+  /**
+   * Returns the y-axis location for the yChip'th chip of the yHost'th host.
+   */
+  getChipYLoc: function(yHost: number, yChip: number): number {
+    return yHost * this._hostGridHeight + HOST_TO_CHIP_MARGIN +
+        yChip * (CHIP_GRID_SIZE + CHIP_TO_CHIP_MARGIN);
+  },
+  /**
+   * Returns the x-axis location for the xNode'th node of the xChip'th chip of
+   * the xHost'th host.
+   */
+  getNodeXLoc: function(xHost: number, xChip: number, xNode: number): number {
+    return this.getChipXLoc(xHost, xChip) + xNode * this._nodeGridWidth;
+  },
+  /**
+   * Returns the location for each host in the system.
+   */
+  createHostData: function(
+      hostXDim: number, hostYDim: number): Array<HostData> {
+    let hostData = [];
+    for (let i = 0; i < hostXDim; i++) {
+      for (let j = 0; j < hostYDim; j++) {
+        hostData.push({xdim: i, ydim: j});
+      }
+    }
+    return hostData;
+  },
+  /**
+   * Draw the labels on x-axis and y-axis.
+   */
+  drawLabels: function(svg: any, xdims: number[], ydims: number[]) {
+    let parent = this;
+
+    // Draw label on x axis.
+    svg.selectAll('.xLabel')
+       .data(xdims)
+       .enter()
+       .append('text')
+       .text((d) => d)
+       .attr('x',
+             (d, i) => parent.getChipXLoc(
+                 Math.floor(i / parent._hostXStride), i % parent._hostXStride))
+       .attr('y', 0)
+       .style('text-anchor', 'middle')
+       .attr('transform', 'translate(' + CHIP_GRID_SIZE / 2 + ', -6)')
+       .attr('class', 'axis');
+
+    // Draw label on y axis.
+    svg.selectAll('.yLabel')
+       .data(ydims)
+       .enter()
+       .append('text')
+       .text((d) => d)
+       .attr('x', 0)
+       .attr('y',
+             (d, i) => parent.getChipYLoc(
+                 Math.floor(i / HOST_Y_STRIDE), i % HOST_Y_STRIDE))
+       .style('text-anchor', 'middle')
+       .attr('transform', 'translate(-12,' + CHIP_GRID_SIZE / 2 + ')')
+       .attr('class', 'axis');
+  },
+  /**
+   * Draw the UI of host cards.
+   */
+  drawHostCards: function(svg, data, gridWidth: number, gridHeight: number) {
+    let cards = svg.selectAll('.xdim').data(data, (d) => d.xdim);
+    cards.enter()
```

Comment:
Can the `data` ever change? When it changes, do you have to update the existing nodes? If so, this is incorrect.

Reply:
When the data changed, drawTopology will be called, which calls drawHostCards.
cards.merge, and cards.exit().remove should take care of the updates.

Reply:
Currently, the host dimension changes only when run/tool/host changes, which everything on the page refreshes.